---
title: "SIP pipeline for reproducing Kapili et al., 2019: dada2 ASV inference and HTSSIP"
output: html_notebook
---

dada2 pipeline code modified from:
https://benjjneb.github.io/dada2/tutorial.html

Install required packages
```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("dada2")
biocLite('phyloseq')
biocLite("ShortRead")
install.packages("ggplot2")
install.packages("Hmisc")
install.packages("dplyr")
install.packages("tidyverse")
install.packages("HTSSIP")
install.packages("DESeq2")
```

Load required packages
```{r}
library(dada2)
library(ShortRead)
library(ggplot2)
library(phyloseq)
library(Hmisc)
library(HTSSIP)
library(tidyverse)
library(dplyr)
library(DESeq2)
```

Set path and grab filenames
```{r}
path <- "PATH_TO_SIP_FILES" # Replace with location of downloaded sequences
fns <- list.files(path)
fastqs <- fns[grepl(".fastq$", fns)]
fastqs <- sort(fastqs) # Sort ensures forward/reverse reads are in same order

fnFs <- fastqs[grepl("_R1", fastqs)] # Just the forward read files
fnRs <- fastqs[grepl("_R2", fastqs)] # Just the reverse read files

fnFs_WFSIP01 <- fnFs[grepl("-WFSIP01", fnFs)] # Just the forward reads from WFSIP01
fnRs_WFSIP01 <- fnRs[grepl("-WFSIP01", fnRs)] # Just the reverse reads from WFSIP01

fnFs_WFSIP02 <- fnFs[grepl("-WFSIP02", fnFs)] # Just the forward reads from WFSIP02
fnRs_WFSIP02 <- fnRs[grepl("-WFSIP02", fnRs)] # Just the reverse reads from WFSIP02

fnFs_WFSIP03 <- fnFs[grepl("-WFSIP03", fnFs)] # Just the forward reads from WFSIP03
fnRs_WFSIP03 <- fnRs[grepl("-WFSIP03", fnRs)] # Just the reverse reads from WFSIP03


head(fnFs)
head(fnRs)
```

Get sample names, assuming files named as: SAMPLENAME_XXX.fastq
```{r}
sample.names_WFSIP01 <- sapply(strsplit(fnFs_WFSIP01, "_"), `[`, 1)
sample.names_WFSIP02 <- sapply(strsplit(fnFs_WFSIP02, "_"), `[`, 1)
sample.names_WFSIP03 <- sapply(strsplit(fnFs_WFSIP03, "_"), `[`, 1)

sample.names_WFSIP03
```

Specify the full path to the fnFs and fnRs
```{r}
fnFs_WFSIP01 <- file.path(path, fnFs_WFSIP01)
fnRs_WFSIP01 <- file.path(path, fnRs_WFSIP01)

fnFs_WFSIP02 <- file.path(path, fnFs_WFSIP02)
fnRs_WFSIP02 <- file.path(path, fnRs_WFSIP02)

fnFs_WFSIP03 <- file.path(path, fnFs_WFSIP03)
fnRs_WFSIP03 <- file.path(path, fnRs_WFSIP03)

plotQualityProfile(fnFs_WFSIP01[[1]])
plotQualityProfile(fnRs_WFSIP01[[1]])
```

Make directory and filenames for the filtered fastqs
```{r}
filt_path <- file.path(path, "filtered")
if(!file_test("-d", filt_path)) dir.create(filt_path)

filtFs_WFSIP01 <- file.path(filt_path, paste0(sample.names_WFSIP01, "_F_filt.fastq"))
filtRs_WFSIP01 <- file.path(filt_path, paste0(sample.names_WFSIP01, "_R_filt.fastq"))

filtFs_WFSIP02 <- file.path(filt_path, paste0(sample.names_WFSIP02, "_F_filt.fastq"))
filtRs_WFSIP02 <- file.path(filt_path, paste0(sample.names_WFSIP02, "_R_filt.fastq"))

filtFs_WFSIP03 <- file.path(filt_path, paste0(sample.names_WFSIP03, "_F_filt.fastq"))
filtRs_WFSIP03 <- file.path(filt_path, paste0(sample.names_WFSIP03, "_R_filt.fastq"))
```


Filter previously trimmed fastq files
```{r}
#Filter those from run WFSIP01
for(i in seq_along(fnFs_WFSIP01)) {
  fastqPairedFilter(c(fnFs_WFSIP01[i], fnRs_WFSIP01[i]), c(filtFs_WFSIP01[i], filtRs_WFSIP01[i]),
                    truncLen=c(220,220), 
                    maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
                    compress=F, verbose=TRUE)
}

#Filter those from run WFSIP02
for(i in seq_along(fnFs_WFSIP02)) {
  fastqPairedFilter(c(fnFs_WFSIP02[i], fnRs_WFSIP02[i]), c(filtFs_WFSIP02[i], filtRs_WFSIP02[i]),
                    truncLen=c(220,220), 
                    maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
                    compress=F, verbose=TRUE)
}

#Filter those from run WFSIP03
for(i in seq_along(fnFs_WFSIP03)) {
  fastqPairedFilter(c(fnFs_WFSIP03[i], fnRs_WFSIP03[i]), c(filtFs_WFSIP03[i], filtRs_WFSIP03[i]),
                    truncLen=c(220,220), 
                    maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
                    compress=F, verbose=TRUE)
}
```

Make sure vectors only point to written files, in event sample ends up with 0 sequences after filtering
```{r}
fns_filt <- list.files(filt_path)
fastqs_filt <- fns_filt[grepl(".fastq$", fns_filt)]
fastqs_filt <- sort(fastqs_filt)
filtFs <- fastqs_filt[grepl("_F", fastqs_filt)] 
filtRs <- fastqs_filt[grepl("_R", fastqs_filt)]

filtFs_WFSIP01 <- filtFs[grepl("-WFSIP01", filtFs)] 
filtRs_WFSIP01 <- filtRs[grepl("-WFSIP01", filtRs)]
sample.names.filt_WFSIP01 <- sapply(strsplit(filtFs_WFSIP01, "_"), `[`, 1)

filtFs_WFSIP02 <- filtFs[grepl("-WFSIP02", filtFs)] 
filtRs_WFSIP02 <- filtRs[grepl("-WFSIP02", filtRs)]
sample.names.filt_WFSIP02 <- sapply(strsplit(filtFs_WFSIP02, "_"), `[`, 1)

filtFs_WFSIP03 <- filtFs[grepl("-WFSIP03", filtFs)] 
filtRs_WFSIP03 <- filtRs[grepl("-WFSIP03", filtRs)]
sample.names.filt_WFSIP03 <- sapply(strsplit(filtFs_WFSIP03, "_"), `[`, 1)
```

Dereplicate sequence files
```{r}
setwd(filt_path)
derepFs_WFSIP01 <- derepFastq(filtFs_WFSIP01, verbose=TRUE)
derepRs_WFSIP01 <- derepFastq(filtRs_WFSIP01, verbose=TRUE)

derepFs_WFSIP02 <- derepFastq(filtFs_WFSIP02, verbose=TRUE)
derepRs_WFSIP02 <- derepFastq(filtRs_WFSIP02, verbose=TRUE)

derepFs_WFSIP03 <- derepFastq(filtFs_WFSIP03, verbose=TRUE)
derepRs_WFSIP03 <- derepFastq(filtRs_WFSIP03, verbose=TRUE)
```

Name the derep-class objects by the sample names
```{r}
names(derepFs_WFSIP01) <- sample.names.filt_WFSIP01
names(derepRs_WFSIP01) <- sample.names.filt_WFSIP01

names(derepFs_WFSIP02) <- sample.names.filt_WFSIP02
names(derepRs_WFSIP02) <- sample.names.filt_WFSIP02

names(derepFs_WFSIP03) <- sample.names.filt_WFSIP03
names(derepRs_WFSIP03) <- sample.names.filt_WFSIP03
```

Determine error rates for each run separately
```{r}
dadaFs.lrn_WFSIP01 <- dada(derepFs_WFSIP01, err=NULL, selfConsist = TRUE, multithread=TRUE)
errF_WFSIP01 <- dadaFs.lrn_WFSIP01[[1]]$err_out

dadaRs.lrn_WFSIP01 <- dada(derepRs_WFSIP01, err=NULL, selfConsist = TRUE, multithread=TRUE)
errR_WFSIP01 <- dadaRs.lrn_WFSIP01[[1]]$err_out

dadaFs.lrn_WFSIP02 <- dada(derepFs_WFSIP02, err=NULL, selfConsist = TRUE, multithread=TRUE)
errF_WFSIP02 <- dadaFs.lrn_WFSIP02[[1]]$err_out

dadaRs.lrn_WFSIP02 <- dada(derepRs_WFSIP02, err=NULL, selfConsist = TRUE, multithread=TRUE)
errR_WFSIP02 <- dadaRs.lrn_WFSIP02[[1]]$err_out

dadaFs.lrn_WFSIP03 <- dada(derepFs_WFSIP03, err=NULL, selfConsist = TRUE, multithread=TRUE)
errF_WFSIP03 <- dadaFs.lrn_WFSIP03[[1]]$err_out

dadaRs.lrn_WFSIP03 <- dada(derepRs_WFSIP03, err=NULL, selfConsist = TRUE, multithread=TRUE)
errR_WFSIP03 <- dadaRs.lrn_WFSIP03[[1]]$err_out

plotErrors(dadaFs.lrn_WFSIP01[[1]], nominalQ=TRUE)
```

Infer the sequence variants in each sample
```{r}
dadaFs_WFSIP01 <- dada(derepFs_WFSIP01, err=errF_WFSIP01, multithread=TRUE, pool=TRUE)
dadaRs_WFSIP01 <- dada(derepRs_WFSIP01, err=errR_WFSIP01, multithread=TRUE, pool=TRUE)

dadaFs_WFSIP02 <- dada(derepFs_WFSIP02, err=errF_WFSIP02, multithread=TRUE, pool=TRUE)
dadaRs_WFSIP02 <- dada(derepRs_WFSIP02, err=errR_WFSIP02, multithread=TRUE, pool=TRUE)

dadaFs_WFSIP03 <- dada(derepFs_WFSIP03, err=errF_WFSIP03, multithread=TRUE, pool=TRUE)
dadaRs_WFSIP03 <- dada(derepRs_WFSIP03, err=errR_WFSIP03, multithread=TRUE, pool=TRUE)
```

Merge paired-end reads
```{r}
mergers_WFSIP01 <- mergePairs(dadaFs_WFSIP01, derepFs_WFSIP01, dadaRs_WFSIP01, derepRs_WFSIP01, verbose=TRUE)
mergers_WFSIP02 <- mergePairs(dadaFs_WFSIP02, derepFs_WFSIP02, dadaRs_WFSIP02, derepRs_WFSIP02, verbose=TRUE)
mergers_WFSIP03 <- mergePairs(dadaFs_WFSIP03, derepFs_WFSIP03, dadaRs_WFSIP03, derepRs_WFSIP03, verbose=TRUE)
```

Make an 'OTU' (ASV) table
```{r}
seqtab_WFSIP01 <- makeSequenceTable(mergers_WFSIP01)
seqtab_WFSIP02 <- makeSequenceTable(mergers_WFSIP02)
seqtab_WFSIP03 <- makeSequenceTable(mergers_WFSIP03)

seqtab_WFSIP <- mergeSequenceTables(seqtab_WFSIP01,seqtab_WFSIP02,seqtab_WFSIP03)
dim(seqtab_WFSIP)
```

Inspect distribution of sequence lengths
```{r}
table(nchar(getSequences(seqtab_WFSIP)))
seq.lengths<-nchar(getSequences(seqtab_WFSIP))
head(seq.lengths)

max(nchar(getSequences(seqtab_WFSIP)))
min(nchar(getSequences(seqtab_WFSIP)))

hist.breaks<-seq(210,430,by=10)

hist(seq.lengths,breaks=hist.breaks,xaxt="n")
  axis(1,hist.breaks)

hist(seq.lengths,breaks=hist.breaks,xaxt="n",ylim=c(0,100))
  axis(1,hist.breaks)

  describe(seq.lengths)
summary(seq.lengths)
```

Filter sequences to retain just those between 368-378 bp
```{r}
seqtab2_WFSIP <- seqtab_WFSIP[,nchar(colnames(seqtab_WFSIP)) %in% seq(368,378)]
```

Remove chimeric sequences
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab2_WFSIP, verbose=TRUE)
dim(seqtab.nochim) #how many sequences passed 
sum(seqtab.nochim)/sum(seqtab2_WFSIP) #percentage of sequences not chimeras

```

Optional: Save the filtered OTU table
```{r}
#saveRDS(seqtab.nochim, "seqtabnochim_20180815_WFSIP.rds")
# If need to Close and re-open
#seqtab.nochim <- readRDS("seqtabnochim_20180815_WFSIP.rds")
```

Assign taxonomy using Silva train set v132
```{r}
taxa <- assignTaxonomy(seqtab.nochim, "~/Desktop/Shared_Folder/silva_nr_v132_train_set.fa")
unname(head(taxa))
```

Create dataframe with sample information
```{r}
otu.seqs<-colnames(seqtab.nochim)
otu.names<-paste("OTU",1:length(otu.seqs),sep=".")
otu.df<-data.frame(otu.names,otu.seqs)
seqtab.otu<-seqtab.nochim
colnames(seqtab.otu)<-otu.df$otu.names
rownames(taxa)<-otu.df$otu.names

samples.out <- rownames(seqtab.otu)
sample <- sapply(strsplit(samples.out, "-"), `[`, 1)
fraction <- sapply(strsplit(samples.out, "-"), `[`, 2)
depth <- sapply(strsplit(samples.out, "-"), `[`, 3)
rep <- sapply(strsplit(samples.out, "-"), `[`, 4)
run <- sapply(strsplit(samples.out, "-"), `[`, 5)

samdf <- data.frame(Sample=sample,Fraction=fraction,Depth=depth,
                      Rep=rep,Run=run)
rownames(samdf) <- samples.out
```

Make phyloseq object that incorporates sample information dataframe (samdf), OTU table (seqtab.otu), taxonomy information (taxa), and sequence data (seqs)
```{r}
seqs <- DNAStringSet(otu.df$otu.seqs)
names(seqs) <- otu.df$otu.names

ps <- phyloseq(otu_table(seqtab.otu, taxa_are_rows=FALSE),  
               tax_table(taxa), sample_data(samdf), refseq(seqs))

ps
```

Optional: Save phyloseq object
```{r}
#saveRDS(ps, "psWFSIP_20180815.rds")
# if you need to Close R and re-open from here
#ps <- readRDS("psWFSIP_20180815.rds")
```

Remove singleton taxa
```{r}
ps3<-prune_taxa(taxa_sums(ps)>1,ps)
ps3
```

Remove Eurkaryotes and Unclassified taxa
```{r}
psPruned = subset_taxa(ps3, Kingdom !="Eukaryota" & Kingdom != "NA" )
psPruned.abund <- transform_sample_counts(psPruned, function(x) 100 * x/sum(x))

plot_bar(psMergePruned,fill="Phylum",facet_grid=~Kingdom) +
  coord_flip() +
  geom_bar(stat="identity")

p <- plot_bar(psMergePruned,  fill = "Phylum") +
  coord_flip() 
p + facet_wrap(~nucleicAcidType) + geom_bar(stat="identity")
```

Append density values and treatment type to each sample
```{r}
# Import a dataframe that has the sample names as the first column
# and the corresponding buoyant density values as the second column
density <- read.table("INSERT_PATH_HERE", sep=",",header=TRUE)
treatment <- read.table("INSERT_PATH_HERE", sep=",", header=TRUE)

# This line matches the density values in your new dataframe
# to the corresponding sample names of your phyloseq object
sample_data(psPruned)$Buoyant_density <- density$Density[match(sample_names(psPruned),density$Sample)]

sample_data(psPruned)$Treatment <- treatment$Treatment[match(sample_names(psPruned),treatment$Sample)]
```

SIP

Load in customized functions for SIP
```{r}
#Changes size factor estimation mode to 'poscounts'
DESeq2_l2fc_bjk <- function (physeq, density_min, density_max, design, l2fc_threshold = 0.25, 
    sparsity_threshold = 0.25, sparsity_apply = "all") 
{
    l2fc_threshold = as.numeric(l2fc_threshold)
    stopifnot(l2fc_threshold >= 0 & l2fc_threshold <= 1)
    sparsity_apply = tolower(sparsity_apply)
    stopifnot(sparsity_apply %in% c("all", "heavy"))
    physeq.md = phyloseq::sample_data(physeq)
    stopifnot(!is.null(physeq.md$Buoyant_density))
    cat("Sparsity threshold:", sparsity_threshold, "\n")
    cat("Density window:", paste(c(density_min, density_max), 
        collapse = "-"), "\n")
    prn = function(x) sum(x > 0) > sparsity_threshold * length(x)
    if (sparsity_apply == "all") {
        physeq = phyloseq::filter_taxa(physeq, prn, TRUE)
    }
    physeq = phyloseq::prune_samples((physeq.md$Buoyant_density >= 
        density_min) & (physeq.md$Buoyant_density <= density_max), 
        physeq)
    physeq = phyloseq::filter_taxa(physeq, function(x) sum(x > 
        0) > 0 * length(x), TRUE)
    if (sparsity_apply == "heavy") {
        physeq = phyloseq::filter_taxa(physeq, prn, TRUE)
    }
    dds = phyloseq::phyloseq_to_deseq2(physeq, design=~Sample)
    dds <- estimateSizeFactors(dds, type='poscount')
    dds <- DESeq2::DESeq(dds, quiet = TRUE, fitType = "local")
    theta = l2fc_threshold
    res = DESeq2::results(dds, independentFiltering = TRUE)
    res$OTU = rownames(res)
    beta = res$log2FoldChange
    betaSE = res$lfcSE
    p = stats::pnorm(beta, theta, betaSE, lower.tail = FALSE)
    res$p = p
    d = data.frame(res[, c("OTU", "log2FoldChange", "p")])
    d$padj = stats::p.adjust(p, method = "BH")
    TT = phyloseq::tax_table(physeq)
    if (!is.null(TT)) {
        TT = as.data.frame(as.matrix(TT))
        TT$OTU = rownames(TT)
        d = dplyr::left_join(d, TT, c("OTU"))
    }
    d$density_min = density_min
    d$density_max = density_max
    d$sparsity_threshold = sparsity_threshold
    d$sparsity_apply = sparsity_apply
    d$l2fc_threshold = l2fc_threshold
    return(d)
}

#Calls the updated DESeq2 function
HRSIP_bjk <- function (physeq, design, density_windows = data.frame(density_min = c(1.7), 
    density_max = c(1.75)), sparsity_threshold = seq(0, 0.3, 
    0.1), sparsity_apply = "all", l2fc_threshold = 0.25, padj_method = "BH", 
    padj_cutoff = NULL, parallel = FALSE) 
{
    if (is.factor(density_windows)) {
        density_windows = as.vector(density_windows)
    }
    if (is.vector(density_windows)) {
        stopifnot(length(density_windows) >= 2)
        density_windows = data.frame(start = c(density_windows[1]), 
            end = c(density_windows[2]))
    }
    stopifnot(all(c("density_min", "density_max") %in% colnames(density_windows)))
    stopifnot(is.numeric(l2fc_threshold))
    stopifnot(is.character(sparsity_apply))
    stopifnot(all(sapply(sparsity_threshold, function(x) x >= 
        0 & x <= 1)) == TRUE)
    stopifnot(is.data.frame(density_windows))
    stopifnot(ncol(density_windows) >= 2)
    density_windows$.LEFT_JOIN_COLUMN = 1
    sparsity_threshold = as.data.frame(sparsity_threshold)
    sparsity_threshold$.LEFT_JOIN_COLUMN = 1
    m = dplyr::left_join(density_windows, sparsity_threshold, 
        c(".LEFT_JOIN_COLUMN"))
    m$.LEFT_JOIN_COLUMN = NULL
    df_l2fc = plyr::mdply(m, DESeq2_l2fc_bjk, physeq = physeq, design = design, 
        l2fc_threshold = l2fc_threshold, sparsity_apply = sparsity_apply, 
        .parallel = parallel)
    mutate_call = lazyeval::interp(~stats::p.adjust(x, method = padj_method), 
        x = as.name("p"))
    dots = stats::setNames(list(mutate_call), "padj")
    df_l2fc = df_l2fc %>% dplyr::group_by_("sparsity_threshold") %>% 
        dplyr::mutate_(.dots = dots) %>% dplyr::ungroup()
    if (!is.null(padj_cutoff)) {
        df_l2fc = filter_l2fc(df_l2fc, padj_cutoff = padj_cutoff)
    }
    return(df_l2fc)
}

filter_l2fc = function(df_l2fc, padj_cutoff=0.1){
  padj_cutoff = as.numeric(padj_cutoff)

  # filter to sparsity thresholds with > number of rej_hypo
  ## init dots
  mutate_call = lazyeval::interp(~ sum(padj<padj_cutoff),
                                 padj = as.name('padj'))
  dots = stats::setNames(list(mutate_call), 'n_rej_hypo')
  ## which sparsity cutoff and BD
  df_l2fc_s = df_l2fc %>%
    # number of rej hypo
    dplyr::group_by_("sparsity_threshold") %>%
    dplyr::summarize_(.dots=dots) %>%
    dplyr::ungroup() %>%
    dplyr::mutate_(rank_n_rej_hypo = "dplyr::row_number(-n_rej_hypo)") %>%
    dplyr::filter_("rank_n_rej_hypo == 1")

  ## best sparsity cutoff
  BEST_SPAR_THRESH = as.numeric(df_l2fc_s[1,'sparsity_threshold'])
  cat('Sparsity threshold with the most rejected hypotheses:', BEST_SPAR_THRESH, '\n')
  ### filtering
  mutate_call = lazyeval::interp(~ x==BEST_SPAR_THRESH,
                                 x=as.name('sparsity_threshold'))
  dots = stats::setNames(list(mutate_call), NA)
  df_l2fc = df_l2fc %>%
    dplyr::filter_(.dots=dots)

  # For each sparsity threshold, selecting OTUs with highest l2fc
  ## filtering OTUs to just density window with the highest l2fc
  df_l2fc = df_l2fc %>%
    dplyr::group_by_("OTU", "sparsity_threshold") %>%
    dplyr::filter_("log2FoldChange == max(log2FoldChange)") %>%
    dplyr::ungroup()

  return(df_l2fc)
}
```

Create phyloseq objects of just the unfractionated samples 
```{r}
# Create a separate phyloseq object just for the unfractionated samples
psUnfrac <- subset_samples(psPruned, Fraction == 'unfrac')

psUnfrac.52 <- subset_samples(psUnfrac, Sample == 'PC6503' | Sample == 'WF52')
psUnfrac.55 <- subset_samples(psUnfrac, Sample == 'PC6503' | Sample == 'WF55')
psUnfrac.83 <- subset_samples(psUnfrac, Sample == 'PC65912' | Sample == 'WF83')
psUnfrac.86 <- subset_samples(psUnfrac, Sample == 'PC65912' | Sample == 'WF86')
```

```{r}
#Define data frame to hold treatment/control abundance ratios
df.ratio <- data.frame(rownames=colnames(otu_table(psUnfrac)),
                       WF52 = 'NA',
                       WF55 = 'NA',
                       WF83 = 'NA',
                       WF86 = 'NA')

#Define function to return median-of-ratios normalized counts for each control-treatment pairing
dds.return <- function(x) {
  dds = phyloseq::phyloseq_to_deseq2(x, design=~Sample)
  dds <- estimateSizeFactors(dds, type='poscounts')
  cnts <- as.data.frame(counts(dds, normalized=TRUE))
  return(as.numeric(cnts[,2]/cnts[,1]))
}

#Create grouped phyloseq object of each control-treatment pairing
phylos <- c(psUnfrac.52,psUnfrac.55,psUnfrac.83,psUnfrac.86)

#Apply function to each pairing and store ratios in df.ratio
df.ratio[,2:5] <- lapply(phylos,dds.return)
```

```{r}
# Remove all mock communities, technical negatives, and technical replicates
psSIP.pruned <- subset_samples(psPruned, Buoyant_density != 'NA')
psSIP.pruned <- subset_samples(psSIP.pruned, Rep != 'Rep2')

# Remove Replicate from PC65912 H2
psSIP.pruned <- subset_samples(psSIP.pruned, sample_names(psSIP.pruned) != 'PC65912-H2-9cm12cm-Rep2-WFSIP02')

# Remove WF60 samples that weren't sequenced properly during WFSIP01
psSIP.pruned <- subset_samples(psSIP.pruned, sample_names(psSIP.pruned) != 'WF60-A4-0cm3cm-Rep1-WFSIP01')
psSIP.pruned <- subset_samples(psSIP.pruned, sample_names(psSIP.pruned) != 'WF60-B4-0cm3cm-Rep1-WFSIP01')
psSIP.pruned <- subset_samples(psSIP.pruned, sample_names(psSIP.pruned) != 'WF60-C4-0cm3cm-Rep1-WFSIP01')
psSIP.pruned <- subset_samples(psSIP.pruned, sample_names(psSIP.pruned) != 'WF60-F2-0cm3cm-Rep1-WFSIP01')
psSIP.pruned <- subset_samples(psSIP.pruned, sample_names(psSIP.pruned) != 'WF60-F3-0cm3cm-Rep1-WFSIP01')

# Group control/treatment in a phyloseq object
comps <- 
  "(Treatment == 'Control' & Depth == '${Depth}') | (Treatment == '${Treatment}' & Depth == '${Depth}')"
params <- get_treatment_params(psSIP.pruned, c('Treatment','Depth'),"Treatment != 'Control'")

psSIP.comp <- phyloseq_subset(psSIP.pruned, params, comps)
```

Filtering step 1: Remove OTUs that are 10X more abundant or 10X less abundant in treatments in comparison to controls
```{r}
#Set abundance ratio cutoff
x=10

#Determine which OTUs meet the abundance ratio cutoff
WF52.keep <- df.ratio[which(df.ratio[,2] < x & df.ratio[,2] > 1/x),1]
WF55.keep <- df.ratio[which(df.ratio[,3] < x & df.ratio[,3] > 1/x),1]
WF83.keep <- df.ratio[which(df.ratio[,4] < x & df.ratio[,4] > 1/x),1]
WF86.keep <- df.ratio[which(df.ratio[,5] < x & df.ratio[,5] > 1/x),1]

#Prune OTUs that don't meet the cutoff
psSIP.comp$`(Treatment == 'Control' & Depth == '0cm3cm') | (Treatment == 'CH4+15N2' & Depth == '0cm3cm')`<-  
  prune_taxa(as.vector(WF52.keep), psSIP.comp$`(Treatment == 'Control' & Depth == '0cm3cm') | (Treatment == 'CH4+15N2' & Depth == '0cm3cm')`)

psSIP.comp$`(Treatment == 'Control' & Depth == '0cm3cm') | (Treatment == 'Ar+15N2' & Depth == '0cm3cm')` <- 
  prune_taxa(as.vector(WF55.keep), psSIP.comp$`(Treatment == 'Control' & Depth == '0cm3cm') | (Treatment == 'Ar+15N2' & Depth == '0cm3cm')`)

psSIP.comp$`(Treatment == 'Control' & Depth == '9cm12cm') | (Treatment == 'CH4+15N2' & Depth == '9cm12cm')` <- 
  prune_taxa(as.vector(WF83.keep), psSIP.comp$`(Treatment == 'Control' & Depth == '9cm12cm') | (Treatment == 'CH4+15N2' & Depth == '9cm12cm')`)

psSIP.comp$`(Treatment == 'Control' & Depth == '9cm12cm') | (Treatment == 'Ar+15N2' & Depth == '9cm12cm')` <- 
  prune_taxa(as.vector(WF86.keep), psSIP.comp$`(Treatment == 'Control' & Depth == '9cm12cm') | (Treatment == 'Ar+15N2' & Depth == '9cm12cm')`)

psSIP.comp <- psSIP.comp[-6]
psSIP.comp <- psSIP.comp[-3]
```

Filtering step 2: Remove OTUs that aren't present in at least 3 fractions in both treatment and control
```{r}
#Define function that prunes taxa that are not present in at least x fractions greater than 1.69 g ml-1
frac_prune <- function(x,z) {
  p <- prune_samples(sample_data(x)$'Buoyant_density'>1.69,x)
  control <- otu_table(p)[grep('PC', rownames(otu_table(p))),]
  treat <- otu_table(p)[grep("WF5|WF8", rownames(otu_table(p))),]
  
  keep <- otu_table(p)[,colSums(control != 0) >=z & colSums(treat != 0) >=z]

  t <- prune_taxa(colnames(keep), x)
  return(t)
}


#Prune taxa not present in at least 3 fractions in each control and treatment >1.69 g ml-1
psHTSSIP <- lapply(psSIP.comp,frac_prune,3)
```

Run HTSSIP
```{r}
#Run HTSSIP
windows = data.frame(density_min=c(1.70, 1.7075, 1.715, 1.7225), 
                     density_max=c(1.715, 1.7225, 1.73, 1.7375))

sparsity = c(0.35)

doParallel::registerDoParallel(2)

df_l2fc.WF = plyr::ldply(psHTSSIP, 
                      HRSIP_bjk, 
                      density_windows = windows,
                      design = ~Sample, 
                      padj_cutoff = 0.05,
                      l2fc_threshold = 0.25,
                      sparsity_threshold = sparsity,
                      sparsity_apply = "heavy",
                      .parallel = TRUE)
df_l2fc.WF %>% head(n=3)

df_l2fc.WF %>% .$.id %>% unique

df_l2fc.WF.filt <- df_l2fc.WF %>%
  filter(padj < 0.05) #Keep only those with adjusted p-values < 0.05

df_l2fc.WF.removed <- df_l2fc.WF %>%
  filter(padj > 0.05)
```

```{r}
#Save psPruned
#saveRDS(psPruned , "~/Documents/STANFORD/DEKAS/Cornell_Experiment/psPruned.rds")
```